compiler-rt/lib/ubsan/ubsan_signals_standalone.cpp

```diff
@@ -66,6 +66,11 @@ void InitializeDeadlySignals() {
     return;
   is_initialized = true;
   InitializeSignalInterceptors();
+#if SANITIZER_INTERCEPT_SIGNAL_AND_SIGACTION
+  // REAL(sigaction_symname) is nullptr in a static link. Bail out.
+  if (!REAL(sigaction_symname))
```

Comment:
I thin it should be in inside of internal_sigaction
and it should fallback to "return sigaction(" like GO

Reply:
`internal_sigaction` cannot call `sigaction` (interceptor), as there will be a dead lock. In asan, `AsanInitInternal` (which holds `asan_inited_mutex`) calls `InitializeSignalInterceptors`. If the `sigaction` interceptor is called, `asan_inited_mutex` will be held again -> dead lock.

It seems that testing `REAL(sigaction_symname)` in `ubsan` (and future sanitizers that can be used with static linking) is better?

Reply:
@vitalybuka ^ :)

Reply:
standalone UBSAN ubsan without asan, so it should be fine

Reply:
So as-is this patch introduce interceptor related logic into UBSAN, which would be nice to avoid

```
int internal_sigaction(int signum, const void *act, void *oldact) {
#  if !SANITIZER_GO
  if (&real_sigaction)
    return real_sigaction(signum, act, oldact);
#  endif
  return sigaction(signum, (const struct sigaction *)act,
                   (struct sigaction *)oldact);
}
```
From this code, it already should throughout to `sigaction`, so it does not match the test in the description: 

> with static linking
(-static/-static-pie), the called REAL(sigaction) is null, leading to
an immediate segfault,

Reply:
Actually signal interceptors are there :(

I think the best for `-static` is to avoid code
```
#define SIGNAL_INTERCEPTOR_ENTER() __ubsan::InitializeDeadlySignals()
#include "sanitizer_common/sanitizer_signal_interceptors.inc"
```

maybe in any  UBSAN

`InitializeSignalInterceptors` is only needed to enforce options like `handle_segv=2`
which are not very critical

Reply:
> So as-is this patch introduce interceptor related logic into UBSAN, which would be nice to avoid
> 
> ```
> int internal_sigaction(int signum, const void *act, void *oldact) {
> #  if !SANITIZER_GO
>   if (&real_sigaction)
>     return real_sigaction(signum, act, oldact);
> #  endif
>   return sigaction(signum, (const struct sigaction *)act,
>                    (struct sigaction *)oldact);
> }
> ```
> 
> From this code, it already should throughout to `sigaction`, so it does not match the test in the description:
> 
> > with static linking
> > (-static/-static-pie), the called REAL(sigaction) is null, leading to
> > an immediate segfault,

The description is correct.
`real_sigaction` is defined due to `SANITIZER_INTERCEPT_SIGNAL_AND_SIGACTION`.
`real_sigaction` calls a function pointer, `REAL(sigaction)`, which is null in the static linking case.

> I think the best for -static is to avoid code
> 
>> #define SIGNAL_INTERCEPTOR_ENTER() __ubsan::InitializeDeadlySignals()
>> #include "sanitizer_common/sanitizer_signal_interceptors.inc"

Without the `!REAL(sigaction_symname)` check as done in this PR, just making `SIGNAL_INTERCEPTOR_ENTER` a no-op does not fix the static linking crash.

Yes, `ubsan_signals_standalone.cpp` defines the signal interceptors. ISTM adding a few lines in this file is the cleanest approach, in contract to an alternative that modifies a generic file which is not needed by another sanitizer.